### PR TITLE
[BEAM-5605] Convert all BoundedSources to SplittableDoFns when using beam_fn_api experiment.

### PR DESCRIPTION
**Pull request by lukecwik**
Note that this does not yet cover everything needed for progress reporting or dynamic work rebalancing.

------------------------

Thank you for your contribution! Follow this checklist to help us incorporate your contribution quickly and easily:

 - [ ] [**Choose reviewer(s)**](https://beam.apache.org/contribute/#make-your-change) and mention them in a comment (`R: @username`).
 - [ ] Format the pull request title like `[BEAM-XXX] Fixes bug in ApproximateQuantiles`, where you replace `BEAM-XXX` with the appropriate JIRA issue, if applicable. This will automatically link the pull request to the issue.
 - [ ] If this contribution is large, please file an Apache [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.pdf).

See the [Contributor Guide](https://beam.apache.org/contribute) for more tips on [how to make review process smoother](https://beam.apache.org/contribute/#make-reviewers-job-easier).

Post-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

Lang | SDK | Apex | Dataflow | Flink | Gearpump | Samza | Spark
--- | --- | --- | --- | --- | --- | --- | ---
Go | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Go_VR_Spark/lastCompletedBuild/)
Java | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Apex/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Dataflow/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Gearpump/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Samza/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Spark/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_SparkStructuredStreaming/lastCompletedBuild/)
Python | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python2/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python36/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python37/lastCompletedBuild/) | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Py_VR_Dataflow/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Py_ValCont/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python2_PVR_Flink_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PostCommit_Python35_VR_Flink/lastCompletedBuild/) | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_Python_VR_Spark/lastCompletedBuild/)
XLang | --- | --- | --- | [Build Status](https://builds.apache.org/job/beam_PostCommit_XVR_Flink/lastCompletedBuild/) | --- | --- | ---

Pre-Commit Tests Status (on master branch)
------------------------------------------------------------------------------------------------

--- |Java | Python | Go | Website
--- | --- | --- | --- | ---
Non-portable | [Build Status](https://builds.apache.org/job/beam_PreCommit_Java_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Python_Cron/lastCompletedBuild/)<br>[Build Status](https://builds.apache.org/job/beam_PreCommit_PythonLint_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Go_Cron/lastCompletedBuild/) | [Build Status](https://builds.apache.org/job/beam_PreCommit_Website_Cron/lastCompletedBuild/) 
Portable | --- | [Build Status](https://builds.apache.org/job/beam_PreCommit_Portable_Python_Cron/lastCompletedBuild/) | --- | ---

See [.test-infra/jenkins/README](https://github.com/apache/beam/blob/master/.test-infra/jenkins/README.md) for trigger phrase, status and link of all Jenkins jobs.

### CODE REVIEW

**Comment by lukecwik** [comment]
retest this please

**Comment by lukecwik** [comment]
R: @youngoli @boyuanzz 
CC: @tweise @mxm

**Comment by lukecwik** [comment]
please don't merge as I'm testing this within Google to see if this negatively impacts internal pipelines.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/Read.java [hunk 1]
```diff
@@ -94,6 +106,20 @@ private Bounded(@Nullable String name, BoundedSource<T> source) {
     public final PCollection<T> expand(PBegin input) {
       source.validate();
 
+      if (ExperimentalOptions.hasExperiment(input.getPipeline().getOptions(), "beam_fn_api")) {
```
Comment (mxm) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/Read.java]: Does that mean we can remove https://github.com/apache/beam/blob/e1852ca6af92d61467bfa7dac84e96a2924eefac/runners/portability/java/src/main/java/org/apache/beam/runners/portability/PortableRunner.java#L154 ?

Comment (lukecwik) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/io/Read.java]: We could exclude the override if "beam_fn_api" was used as an experiment.

Comment (mxm) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/io/Read.java]: I thought the only application of this override is with the `beam_fn_api` flag enabled. The legacy translation does not support Impulse.

Comment (lukecwik) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/io/Read.java]: Users/framework will need to ensure that the `beam_fn_api` experiment is always used.

**Comment by lukecwik** [comment]
I'm going to wait on #10702 and update this PR after that goes in since it is necessary to support the new DoFn style parameter passing.

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/Read.java (hunk 1 above)
Comment (lukecwik) [reply sdks/java/core/src/main/java/org/apache/beam/sdk/io/Read.java]: I was able to remove JavaReadViaImpulse. It required fixing up some tests that weren't setting the `beam_fn_api` experiment.

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/JavaReadViaImpulse.java [hunk 2]
```diff
@@ -1,176 +0,0 @@
-/*
```
Comment (youngoli) [inline runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/JavaReadViaImpulse.java]: Just to make sure I'm understanding correctly: This transform was only used for portable pipelines, right? So it isn't needed anymore now that we execute bounded reads by wrapping them in an SDF?

**Comment by lukecwik** [comment]
Run Java PreCommit

File: sdks/java/core/src/main/java/org/apache/beam/sdk/io/Read.java [hunk 3]
```diff
@@ -177,4 +205,128 @@ public void populateDisplayData(DisplayData.Builder builder) {
...
+    @GetRestrictionCoder
+    public Coder<BoundedSource<T>> restrictionCoder() {
+      return SerializableCoder.of(new TypeDescriptor<BoundedSource<T>>() {});
+    }
+
+    /**
+     * A fake restriction tracker which adapts to the {@link BoundedSource} API. The restriction
+     * object is used to advance the underlying source and to "return" the current element.
+     */
+    private static class BoundedSourceAsSDFRestrictionTracker<T>
```
Comment (iemejia) [inline sdks/java/core/src/main/java/org/apache/beam/sdk/io/Read.java]: :clap:

File: runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/JavaReadViaImpulse.java (hunk 2 above)
Comment (lukecwik) [reply runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/JavaReadViaImpulse.java]: Yes

Comment (youngoli) [reply runners/core-construction-java/src/main/java/org/apache/beam/runners/core/construction/JavaReadViaImpulse.java]: Gotcha.

I already approved this, but consider it double-approved.

**Comment by lukecwik** [comment]
Run Java PreCommit

**Comment by lukecwik** [comment]
Run Portable_Python PreCommit

**Comment by lukecwik** [comment]
Run Python2_PVR_Flink PreCommit

**Comment by lukecwik** [comment]
Run Python2_PVR_Flink PreCommit

**Comment by lukecwik** [comment]
Run Java PreCommit

**Comment by lukecwik** [comment]
Run Portable_Python PreCommit

**Comment by lukecwik** [comment]
Run JavaPortabilityApi PreCommit

**Comment by lukecwik** [comment]
Run Java PreCommit

**Comment by TheNeuralBit** [comment]
It looks like this broke PVR_{Spark,Flink}

https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/4092/
https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/4104/
https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/2089/

**Comment by lukecwik** [comment]
> It looks like this broke PVR_{Spark,Flink}
> 
> https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Batch/4092/
> https://builds.apache.org/job/beam_PostCommit_Java_PVR_Flink_Streaming/4104/
> https://builds.apache.org/job/beam_PostCommit_Java_PVR_Spark_Batch/2089/

Taking a look.

**Comment by TheNeuralBit** [comment]
AFAICT Spark_Batch was already flaky, but now its failing consistently, and the Flink ones started failing at the same time.

**Comment by iemejia** [comment]
Apparently the fix of Dynamic Timers for Dataflow broke the ValidatesRunner tests for Flink so this is unrelated. See my post in the mailing list.
Tests are failing since the exact commit for the fix: 7719708a04d5d0eff3048dbd58ac1337889f8ba5
For details on the exception:
https://builds.apache.org/job/beam_PostCommit_Java_ValidatesRunner_Flink/6608/

**Comment by lukecwik** [comment]
@iemejia Your both right in two different things being broken.

**Comment by TheNeuralBit** [comment]
The PVR failures I linked above definitely seem to have started when this PR went in: https://github.com/apache/beam/commit/c1d054faf043677d8700de9ed6eada03cccf55e8

**Comment by TheNeuralBit** [comment]
I filed https://issues.apache.org/jira/browse/BEAM-9317 to track this one.

**Comment by iemejia** [comment]
Oh I see this is for the portable runner. Sorry for the extra noise

**Pull request merged**